### PR TITLE
Calculate and add read time estimate to UI

### PR DESCRIPTION
**Pull request by varun-magesh**
This pull requests estimates the read time for a casebook resource. This read time is added to the UI in (what I think is) a non-obtrusive way.

A future PR will build on this work to display estimates for whole sections or casebooks in the interest of helping professors gauge the volume of their casebook.

2022-07-12-163245_850x254_scrot

### CODE REVIEW

**Comment by lizadaly** [comment]
This rules! I think it looks nice too.

I have a few testing suggestions:

Run through a few (5 or so?) different casebooks + resource types and eyeball the estimates. Do they look reasonable?

Write a little testing program using your local database as a datasource and calculate the reading times for every content node, outputting it to a text file (like a csv). Something like this:

```
for node in ContentNode.objects.all():
  print (f"{node.id),{node.reading_time()}")
```

and get a sense of the distribution of values—are a lot of them `None`? Does any node throw an exception?

This wouldn't be a test you'd commit to the repo, just something to validate against live data since we happen to have it easily accessible.

**Comment by varun-magesh** [comment]
sans_html

In general, things look pretty reasonable --- this is a distribution of 1000 nodes.

One drawback to our current approach is that it requires  HTML tags to be removed from every node's content, which is expensive. I tested out a version which didn't bother stripping HTML (but did apply elisions) so that it could be run with a couple of SQL queries, but this version ended up being very inaccurate quite frequently, so we will probably need to cache these results somewhere or be okay with slightly slower load times on casebook pages.

**Comment by lizadaly** [comment]
Yay graphs!

I think ultimately we'll want them precomputed for a lot of reasons, not the least of which would be to display a total reading length for the whole work. But I think this is a terrific feature and unless you're notably concerned with request load time, we can start with on-demand and then maybe add a database field that gets updated any time the content is saved.

**Review by lizadaly — APPROVED** [review]

**Comment by varun-magesh** [comment]
Ultimately, this PR shows an estimated reading time + num links for sections and individual resources. It does not make casebook-level estimates in an effort to trial the feature before deciding on how to expand it best. However, it does provide sufficient backend work to make any display pretty easy.

**Comment by varun-magesh** [comment]
2022-07-14-113128_1415x868_scrot


note: I, too, was skeptical of an entire section of this book only taking 20 minutes to read. However, it turns out that most of the cases within are heavily elided, down to just a paragraph or two, so this number makes sense.

File: web/main/templates/includes/breadcrumbs.html [hunk 1]
```diff
@@ -11,5 +12,13 @@
         <span class="separator">.</span>
       {% endif %}
     {% endfor %}
+    {% if content.reading_time %}
+      <div class="reading-time">
+        {{ content.reading_time | humanize_minutes }}
+        {% if content.num_links %}
+          + {{ content.num_links }} links
```
Comment (lizadaly) [inline web/main/templates/includes/breadcrumbs.html]: What if it's one link? :)

**Comment by lizadaly** [comment]
I like this as an incremental approach and it's nice that you've done enough discovery already to know where to take it next, and have that be easy to implement!

File: web/main/templates/includes/breadcrumbs.html (hunk 1 above)
Comment (varun-magesh) [reply web/main/templates/includes/breadcrumbs.html]: good catch! let me push a fix

**Comment by varun-magesh** [comment]
I'd also appreciate thoughts on the DB persistence approach --- I don't have much experience here, I know my solution works but I'm not sure that it's the paradigmatically correct one.

relevant changes should be in: https://github.com/harvard-lil/h2o/pull/1621/commits/d9eb7d0a6a04d9002f4b7c2882b69060bc3f5730

and https://github.com/harvard-lil/h2o/pull/1621/commits/77c4728f9ee67604d55218d30f40416b953a5f23

File: web/main/templates/includes/breadcrumbs.html (hunk 1 above)
Comment (varun-magesh) [reply web/main/templates/includes/breadcrumbs.html]: should be done!

**Comment by varun-magesh** [comment]
No, the reading time won't be updated on edit --- but it is updated whenever it is needed. It's usually not too expensive --- maybe less than.25s or so on my machine for even the longest content nodes. I will certainly make the change, but I'm not sure I fully understand the benefits of that approach.

**Comment by varun-magesh** [comment]
Hmm, just realized that updating reading time works correctly for text blocks but not eliding legal docs. ;(

**Comment by lizadaly** [comment]
I see that you null out the `reading_length` value on save, I get it now!

**Review by lizadaly — APPROVED** [review]

File: web/main/templates/includes/breadcrumbs.html [hunk 2]
```diff
@@ -11,5 +12,15 @@
         <span class="separator">.</span>
       {% endif %}
     {% endfor %}
+    {% if content.reading_time %}
+      <div class="reading-time {% if editing %}reading-time-tooltip{% endif %}" title="Read time estimate is visible to readers">
+        {{ content.reading_time | humanize_minutes }}
+        {% if content.num_links %}
+        + {{ content.num_links }} link{% if content.num_links > 1 %}s{% endif %}
```
Comment (lizadaly) [inline web/main/templates/includes/breadcrumbs.html]: Django has a built-in for this!

https://docs.djangoproject.com/en/3.2/ref/templates/builtins/#pluralize

(Though honestly it's barely fewer characters.)

**Comment by varun-magesh** [comment]
@lizadaly pushed changes from our conversation yesterday --- this should be ready to go if everything looks good to you!

**Comment by lizadaly** [comment]
Huzzah!

**Pull request merged**